Source/ASDisplayNode.mm

```diff
@@ -438,36 +438,38 @@ - (void)onDidLoad:(ASDisplayNodeDidLoadBlock)body
 - (void)asyncTraitCollectionDidChangeWithPreviousTraitCollection:(ASPrimitiveTraitCollection)previousTraitCollection
 {
   if (@available(iOS 13.0, *)) {
-    __instanceLock__.lock();
-    if (self.primitiveTraitCollection.userInterfaceStyle != previousTraitCollection.userInterfaceStyle) {
-      // When changing between light and dark mode, often the entire node needs to re-render.
-      // This change doesn't happen frequently so it's fairly safe to render nodes again
-      BOOL needsClippingCornerUpdate = NO;
+    // When changing between light and dark mode, often the entire node needs to re-render.
+    // This change doesn't happen frequently so it's fairly safe to render nodes again
+    ASPrimitiveTraitCollection primitiveTraitCollection = self.primitiveTraitCollection;
+    if (primitiveTraitCollection.userInterfaceStyle != previousTraitCollection.userInterfaceStyle) {
+      __instanceLock__.lock();
       CGFloat cornerRadius = _cornerRadius;
       ASCornerRoundingType cornerRoundingType = _cornerRoundingType;
       UIColor *backgroundColor = _backgroundColor;
-      if (_loaded(self)) {
-        CGColorRef cgBackgroundColor = backgroundColor.CGColor;
-        if (!CGColorEqualToColor(_layer.backgroundColor, cgBackgroundColor)) {
-          // Background colors do not dynamically update for layer backed nodes since they utilize CGColorRef
-          // instead of UIColor. Non layer backed node also receive color to the layer (see [_ASPendingState -applyToView:withSpecialPropertiesHandling:]).
-          // We utilize the _backgroundColor instance variable to track the full dynamic color
-          // and apply any changes here when trait collection updates occur.
-          _layer.backgroundColor = cgBackgroundColor;
-        }
-        // If we have clipping corners, re-render the clipping corner layer upon user interface style change
-        if (cornerRoundingType == ASCornerRoundingTypeClipping && cornerRadius > 0.0f) {
-          needsClippingCornerUpdate = YES;
-        }
-      }
       __instanceLock__.unlock();
-      if (needsClippingCornerUpdate) {
-        [self _updateClipCornerLayerContentsWithRadius:cornerRadius backgroundColor:backgroundColor];
+      if (_loaded(self)) {
+        UITraitCollection *traitCollection = ASPrimitiveTraitCollectionToUITraitCollection(primitiveTraitCollection);
+        UIColor *resolvedBackgroundColor = [backgroundColor resolvedColorWithTraitCollection:traitCollection];
+        CGColorRef cgBackgroundColor = resolvedBackgroundColor.CGColor;
+        // traitCollection change might come from layouting, so it is not guaranteed
+        // that we are on the main thread
+        ASPerformBlockOnMainThread(^{
```

Comment:
probably this is not so necessary. May be accessing `layer` outside of the `__instanceLock__ ` will be sufficient.

Reply:
Changing the background color of the layer on the main thread is actually necessary.

One concern I have with this proposal is that, since we get `cgBackgroundColor`, `cornerRoundingType` and `cornerRadius` on a different thread, they might be staled by the time this main thread block is executed. We'll then apply the wrong color and/or radius. I think what we should do is:

- On background thread, check if the trait collection has changed.
- If it's changed, schedule a block to main thread immediately.
- On main thread, grab the latest properties and update the layer if need to.

Reply:
That’s might be problematic cause we might lock both threads (that what Is happenIng now)

Reply:
Not sure I understand you correctly. We wouldn't lock both threads at the same time. On the background thread, we would release the lock before scheduling a block on main thread and grab the lock again on main thread when the block is executed.

Reply:
what is happening now (in my understanding): 
1. we acquire lock on main thread (animation)
2. we acquire lock on background thread (update collection view)
3. seems getting `CALayer` property  somehow locks main thread (unsure)